contracts/nft/erc721m/ERC721CMInitializableRedeemable.sol

```diff
@@ -0,0 +1,636 @@
+    }
+
+    /*==============================================================
+    =                             MODIFIERS                        =
+    ==============================================================*/
+
+    /// @notice Modifier to check if the contract is mintable
+    modifier canMint() {
+        if (!_mintable) revert NotMintable();
+        _;
+    }
+
+    /// @notice Modifier to check if the total supply is enough
+    /// @param qty The quantity to mint
+    modifier hasSupply(uint256 qty) {
+        if (totalSupply() + qty > _maxMintableSupply) revert NoSupplyLeft();
+        _;
+    }
+
+    /*==============================================================
+    =                     PUBLIC WRITE METHODS                     =
+    ==============================================================*/
+
+    /// @notice Mints tokens for the caller
+    /// @param qty The quantity to mint
+    /// @param limit The minting limit for the caller (used in merkle proofs)
+    /// @param proof The merkle proof for allowlist minting
+    /// @param timestamp The timestamp for the minting action (used in cosigning)
+    /// @param signature The cosigner's signature
+    function mint(uint32 qty, uint32 limit, bytes32[] calldata proof, uint256 timestamp, bytes calldata signature)
+        external
+        payable
+        virtual
+        nonReentrant
+    {
+        _mintInternal(qty, msg.sender, limit, proof, timestamp, signature);
+    }
+
+    /// @notice Allows authorized minters to mint tokens for a specified address
+    /// @param to The address to mint tokens for
+    /// @param qty The quantity to mint
+    /// @param limit The minting limit for the recipient (used in merkle proofs)
+    /// @param proof The merkle proof for allowlist minting
+    /// @param timestamp The timestamp for the minting action (used in cosigning)
+    /// @param signature The cosigner's signature
+    function authorizedMint(
+        uint32 qty,
+        address to,
+        uint32 limit,
+        bytes32[] calldata proof,
+        uint256 timestamp,
+        bytes calldata signature
+    ) external payable onlyAuthorizedMinter {
+        _mintInternal(qty, to, limit, proof, timestamp, signature);
+    }
+
+    /// @notice Burns a specific token.
+    /// @dev Only callable by the token owner or an approved operator. The token must exist.
+    /// @param tokenId The ID of the token to burn.
+    function burn(uint256 tokenId) external {
+        _burn(tokenId, true);
+    }
+
+    /// @notice Burns a list of tokens.
+    /// @dev Only callable by the REDEEMER_ROLE, no approval required. The token must exist.
+    /// @param from The owners of the tokens to burn.
+    /// @param tokenIds The IDs of the tokens to burn.
+    function redeem(address[] memory from, uint256[] memory tokenIds) external onlyAuthorizedRedeemer {
+        for (uint256 i = 0; i < tokenIds.length; i++) {
+            if (ownerOf(tokenIds[i]) != from[i]) {
```

Comment:
If the `from` is coming from off chain is this a fiat-based policy? What I mean is since the `authorizedRedeemer` can put an arbitrary list of addresses in the `from` array and the owner of the token is public data what is the enforcement here?

Reply:
so a little background on the workflow. the way we do our physical redemptions is to handle the burn / redemption when the Shopify checkout is completed (order confirmed, inventory is deducted, etc.). Our relayer (we use https://syndicate.io) submits the txn from the shopify webhook callback.

the thinking behind passing in the `from address` for each token id is it adds an extra level of safety in case for some reason those tokens are owned by someone new. 

It just makes me feel more comfortable since these transactions aren't being submitted directly by the user per say.

Reply:
Gotcha gotcha, I'm not sure your opsec/appetite for risk but one suggestion you could try is to rely on [EIP712](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-712.md) for signatures proving ownership. I have some experience with Shopify apps but I don't know their crypto checkout flow (or how your products integrate). The user would need to have their wallet handy that owns the NFT.

For right now, anyone with that key could burn all of the assets without restriction.